### PR TITLE
Composite key for Primitives

### PR DESCRIPTION
**Pull request by ikkoham**
### Summary

I opened this PR for discussion.

=> Conclusion
> We don’t think this is the final solution, but it will certainly be better than the current one (less collision).  I would like to include this mechanism for 0.22, leaving the improvement of better proposals and QuantumCircuit objects as a future issue. 

fixes https://github.com/Qiskit/qiskit-terra/issues/8725

### Details and comments

Alternatives:

1. Use pickle
3. Use qpy 

pickle is a bit faster than qpy but serialization time is unstable...

image

### CODE REVIEW

**Comment by 1ucian0** [comment]
It seems to me that you are trying to implement some sort of `QuantumCircuit.__hash__`. Is this correct?

**Comment by ikkoham** [comment]
True, it's similar, but I don't want to have a hash in the mutable QuantumCircuit, and I believe there is collision for malicious cases in my proposal. Instead of providing hash as an official API, I would like to use it in this way for internal implementation.
For normal hash, object id is not necessary, but for our use case, I think this would be good.

**Comment by ikkoham** [comment]
To fix #8725, let's merge this PR ASAP.

**Comment by t-imamichi** [comment]
It might be good to have more attributes. The following example generates the same value though the number of qubits and classical bits are different.

```python
from qiskit import QuantumCircuit
from qiskit.primitives.utils import _circuit_key

def foo(n):
    qc = QuantumCircuit(n, n, name='foo')
    return qc

for i in range(5):
    print(_circuit_key(foo(i)))
```

output
```
(4367160800, 'foo', 0, 0, None)
(4367160800, 'foo', 0, 0, None)
(4367160800, 'foo', 0, 0, None)
(4367160800, 'foo', 0, 0, None)
(4367160800, 'foo', 0, 0, None)
(4367160800, 'foo', 0, 0, None)
(4367160800, 'foo', 0, 0, None)
(4367160800, 'foo', 0, 0, None)
(4367160800, 'foo', 0, 0, None)
(4367160800, 'foo', 0, 0, None)
```

File: qiskit/primitives/utils.py
Comment (t-imamichi) [inline qiskit/primitives/utils.py:116]: Could you write a short description of this function though it's private?  We may need it for maintenance.
It might be also good to write that we can remove it when `QuantumCircuit.__hash__` is introduced.

**Comment by t-imamichi** [comment]
Another collision example

```python
from qiskit import QuantumCircuit
from qiskit.primitives.utils import _circuit_key

def foo(n):
    qc = QuantumCircuit(1, 1, name='foo')
    qc.ry(n, 0)
    return qc

for i in range(5):
    print(_circuit_key(foo(i)))
```

output
```
(4378819168, 'foo', 1, 0, None)
(4378819168, 'foo', 1, 0, None)
(4378819168, 'foo', 1, 0, None)
(4378819168, 'foo', 1, 0, None)
(4378819168, 'foo', 1, 0, None)
```

**Comment by t-imamichi** [comment]
We perhaps have to enumerate `QuantumCircuit.data` as follows. Note that it does not take care of pulse schedule.

```python
from qiskit import QuantumCircuit
from qiskit.primitives.utils import _circuit_key

def to_tuple(qc):
    lst = []
    for inst in qc.data:
        op = inst.operation
        lst.append(((op.name, op.num_qubits, op.num_clbits, tuple(op.params)), inst.qubits))
    return tuple(lst)

def foo(n):
    qc = QuantumCircuit(1, 1, name='foo')
    qc.ry(n, 0)
    tpl = to_tuple(qc)
    print('foo', n)
    print(tpl, hash(tpl))
    return qc

for i in range(5):
    print(_circuit_key(foo(i)))
```

output:
```
foo 0
((('ry', 1, 0, (0,)), (Qubit(QuantumRegister(1, 'q'), 0),)),) -348870723359866815
(4559362368, 'foo', 1, 0, None)
foo 1
((('ry', 1, 0, (1,)), (Qubit(QuantumRegister(1, 'q'), 0),)),) -260950897317592432
(4559362368, 'foo', 1, 0, None)
foo 2
((('ry', 1, 0, (2,)), (Qubit(QuantumRegister(1, 'q'), 0),)),) -7121263129174142289
(4559362368, 'foo', 1, 0, None)
foo 3
((('ry', 1, 0, (3,)), (Qubit(QuantumRegister(1, 'q'), 0),)),) 7158846673307123618
(4559362368, 'foo', 1, 0, None)
foo 4
((('ry', 1, 0, (4,)), (Qubit(QuantumRegister(1, 'q'), 0),)),) 4696891824434941889
(4559362368, 'foo', 1, 0, None)
```

**Comment by ikkoham** [comment]
pulse schedule is covered by `_op_start_times`

**Comment by t-imamichi** [comment]
We need some test cases of Sampler to cover #8725 and pulse schedule, for example, https://github.com/Qiskit/qiskit-terra/pull/8604#issuecomment-1243187284.

**Comment by t-imamichi** [comment]
If we use `QuantumCircuit.data`, can we exclude `id(QuantumCircuit)` from the key? Do we still need it to avoid collision?

**Comment by t-imamichi** [comment]
It may also good take a plot of runtime of key generation (like https://github.com/Qiskit/qiskit-terra/pull/8604#issue-1348699937) because it enumerates `QuantumCircuit.data` and it is likely to take more time than before. I hope the current PR is still faster than serialization.

**Comment by ikkoham** [comment]
I updated the benchmark of my first comment.

File: qiskit/primitives/utils.py
Comment (t-imamichi) [reply qiskit/primitives/utils.py:137]: sounds good to use `hash`

Comment (ikkoham) [reply qiskit/primitives/utils.py:137]: But hash may collide, i.e. hash(-1) == hash(-2)

Comment (t-imamichi) [reply qiskit/primitives/utils.py:137]: I see. If we directly use a hash key, it may collide... The collision results in a wrong result as #8725. So, we have to use the tuple instead.

**Comment by pedrorrivero** [comment]
I also wonder whether `circuit.name` should be included in the hash. I would agree that it should be in `__hash__` but probably not for indexing circuits here, since it is not functional.

In my humble opinion, this `circuit_key` should be based exclusively on functional elements within `QuantumCircuit`, having non-functional elements added afterwards. for instance by:
```python
def __hash__(self):
    key = _circuit_key(self)
    return hash((self.name, self.metadata, key))
```

File: qiskit/primitives/utils.py
Comment (pedrorrivero) [reply qiskit/primitives/utils.py:137]: Hash collisions are a possibility but very rare. No reasonable hash function is collision-free, but there even are cryptographically secure hashes. The largest the tuple, the less likely a collision.

Notice that these collisions can also happen in dictionary keys, and yet python rolls with it (which is a testament as to how unlikely collisions are). If you use a tuple as a dictionary key, internally python will call `hash()` anyway.

To be fair this is sometimes applied in tandem with `__eq__` so that equality can be checked after matching the hash (these are more advanced implementations of `Mapping`).

We could generate a bunch of random circuits and check for collisions under this paradigm to validate this.

Comment (jakelishman) [reply qiskit/primitives/utils.py:137]: The intent here is to use it as a dictionary key, so you should use the tuple, not the hash of the tuple.  Let Python handle the hash collisions - you'll likely get it wrong if you do it yourself, and CPython's dict implementation is _extremely_ well battle-tested.  Effectively, hash collisions do happen frequently - Python's hash isn't cryptographic (hashes of <61-bit integers are just themselves), and hash-table implementations don't use the full range of the hash otherwise every table would require nearly the entire pointer space of a 64-bit system for the buckets alone.  CPython does it by just taking the x lowest bits (depending on how many buckets it's got allocated).

Since CPython non-empty dicts start with 8 buckets and resize when they exceed 2/3 load factor, with completely random hashes, you can fit 5 items in the collection before it expands.  If the buckets are assigned to completely randomly, you still expect to get a hash collision in 5-element dictionaries $1 - \frac{8! / (8-5)!}{8^5} \approx \frac45$ of the time.

Comment (ikkoham) [reply qiskit/primitives/utils.py:137]: Python uses open addressing to resolve hash collision. This means that key is used to check the given objects are same or not. If key is hash value, we will not be able to avoid this collision. The example I gave is a rare case, but there are actually more collisions. ( It's natural because the hash will lose information.) Id collision exsists, it will be considered to be the same as the previous circuit, and the wrong value will be returned.

File: qiskit/primitives/utils.py
Comment (ikkoham) [inline qiskit/primitives/utils.py:129]:
```suggestion
```

**Comment by t-imamichi** [comment]
Is it possible to make a unit test of pulse schedule?

File: qiskit/primitives/utils.py
Comment (pedrorrivero) [reply qiskit/primitives/utils.py]: That makes sense! Thanks for the detailed explanation 😄

**Comment by t-imamichi** [comment]
Thanks. I'm not familiar with pulse scheduling, but the test looks good to me. I will approve it.

I noticed a tricky case with custom gate. We can make a custom gate with the same name and can cheat this PR.
The following example uses `ry(n)` with different `n` value, but it is enclosed in a custom gate.
So, we have to apply `decompose` or `transpile` to unroll the custom gate.

```python
from qiskit import QuantumCircuit
from qiskit.primitives.utils import _circuit_key

def foo(n):
    circ = QuantumCircuit(1, name='custom')
    circ.ry(n, 0)
    inst = circ.to_instruction()
    qc = QuantumCircuit(1, 1, name='foo')
    qc.append(inst, [0])
    print('foo', n)
    print(qc)
    print(qc.decompose())
    return qc

for i in range(5):
    key = _circuit_key(foo(i))
    print(hash(key), key)
```

output
```
foo 0
     ┌────────┐
  q: ┤ custom ├
     └────────┘
c: 1/══════════

     ┌───────┐
  q: ┤ Ry(0) ├
     └───────┘
c: 1/═════════

-8321684804938097808 (1, 1, 0, (((Qubit(QuantumRegister(1, 'q'), 0),), (), 'custom', ()),), None)
foo 1
     ┌────────┐
  q: ┤ custom ├
     └────────┘
c: 1/══════════

     ┌───────┐
  q: ┤ Ry(1) ├
     └───────┘
c: 1/═════════

-8321684804938097808 (1, 1, 0, (((Qubit(QuantumRegister(1, 'q'), 0),), (), 'custom', ()),), None)
foo 2
     ┌────────┐
  q: ┤ custom ├
     └────────┘
c: 1/══════════

     ┌───────┐
  q: ┤ Ry(2) ├
     └───────┘
c: 1/═════════

-8321684804938097808 (1, 1, 0, (((Qubit(QuantumRegister(1, 'q'), 0),), (), 'custom', ()),), None)
foo 3
     ┌────────┐
  q: ┤ custom ├
     └────────┘
c: 1/══════════

     ┌───────┐
  q: ┤ Ry(3) ├
     └───────┘
c: 1/═════════

-8321684804938097808 (1, 1, 0, (((Qubit(QuantumRegister(1, 'q'), 0),), (), 'custom', ()),), None)
foo 4
     ┌────────┐
  q: ┤ custom ├
     └────────┘
c: 1/══════════

     ┌───────┐
  q: ┤ Ry(4) ├
     └───────┘
c: 1/═════════

-8321684804938097808 (1, 1, 0, (((Qubit(QuantumRegister(1, 'q'), 0),), (), 'custom', ()),), None)
```

**Review by t-imamichi — DISMISSED** [review]

**Comment by ikkoham** [comment]
As my first comment,

> it will certainly be better than the current one (less collision).

The assumption is that this PR is better than the existing one. In offline discussion, this PR does not give a permanent solution nor a hash of QuantumCircuit.

This PR should wait for @jyu00's review, but we merge this once since this is blocking the development of applications. It may be changed in the future, including with or without id.

**Review by t-imamichi — APPROVED** [review]

**Comment by t-imamichi** [comment]
Sounds good. Thank you for the update.

**Pull request merged**
